packages/belos/tpetra/test/TFQMR/test_tfqmr_diag.cpp

```diff
@@ -0,0 +1,428 @@
+  int verbosity = Belos::Errors + Belos::Warnings;
+  if (print)
+    verbosity += Belos::TimingDetails + Belos::StatusTestDetails;
+
+  pList = Teuchos::rcp( new Teuchos::ParameterList );
+  pList->set( "Maximum Iterations", max_iter );
+  pList->set( "Convergence Tolerance", tol );
+  pList->set( "Verbosity", verbosity );
+
+  pBelos = Teuchos::rcp( new TFQMRSolMgr<ST,MV,OP>(pProb, pList) );
+}
+
+template<class OP, class ST, class MP, class MV>
+void Iterative_Inverse_Operator<OP, ST, MP, MV>::operator () (const MV &b, MV &x)
+{
+  int pid = pComm->getRank();
+
+  // Initialize the solution to zero
+  x.putScalar( 0.0 );
+
+  // Reset the solver, problem, and status test for next solve (HKT)
+  pProb->setProblem( Teuchos::rcp(&x, false), Teuchos::rcp(&b, false) );
+
+  timer.start();
+  Belos::ReturnType ret = pBelos->solve();
+  timer.stop();
+
+  if (pid == 0 && print) {
+    if (ret == Belos::Converged)
+    {
+      std::cout << std::endl << "pid[" << pid << "] TFQMR converged" << std::endl;
+      std::cout << "Solution time: " << timer.totalElapsedTime() << std::endl;
+
+    }
+    else
+      std::cout << std::endl << "pid[" << pid << "] TFQMR did not converge" << std::endl;
+  }
+}
+
+//************************************************************************************************
+//************************************************************************************************
+
+template<class ScalarType>
+int run(int argc, char *argv[])
+{
+  // Get default Tpetra template types
+  using ST = typename Tpetra::MultiVector<ScalarType>::scalar_type;
+  using LO = typename Tpetra::MultiVector<>::local_ordinal_type;
+  using GO = typename Tpetra::MultiVector<>::global_ordinal_type;
+  using NT = typename Tpetra::MultiVector<>::node_type;
+
+  // Init Tpetra types
+  using OP = typename Tpetra::Operator<ST,LO,GO,NT>;
+  using MV = typename Tpetra::MultiVector<ST,LO,GO,NT>;
+  using MP = typename Tpetra::Map<LO,GO,NT>;
+
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+
+  Teuchos::GlobalMPISession session(&argc, &argv, NULL);
+  RCP<const Teuchos::Comm<int> > Comm = Tpetra::getDefaultComm();
+  int pid = rank(*Comm);
+
+  bool verbose = false;
+  bool success = true;
+
+  ST tol = sqrt(std::numeric_limits<ST>::epsilon());
+
+  try {
+    int n(10);
+    int numRHS=1;
+
+    RCP<MP> map = RCP(new MP(n, 0, Comm));
+    MV X(map, numRHS), Y(map, numRHS);
+
+    X.putScalar(1.0);
+
+    // Inner computes inv(D2)*y
+    RCP<Diagonal_Operator_2<ST,MV>> D2 = rcp(new Diagonal_Operator_2<ST,MV>(n, map->getMinGlobalIndex(), 1.0));
+    Iterative_Inverse_Operator<OP, ST, MP, MV> A2(n, 1, D2, "Belos (inv(D2))", true);
+
+    // should return x=(1, 1/2, 1/3, ..., 1/10)
+    A2(X,Y);
+
+    if (pid==0) {
+      std::cout << "Vector Y should have all entries [1, 1/2, 1/3, ..., 1/10]" << std::endl;
+    }
+    Y.print(std::cout);
+
+    // Inner computes inv(D)*x
+    RCP<Diagonal_Operator<ST,MV>> D = Teuchos::rcp(new Diagonal_Operator<ST,MV>(n, 4.0));
```

Comment:
you might just say rcp and RCP (as there are using statements): the Teuchos:: prefix is not necessary